fs/xattr.c

```diff
@@ -250,8 +266,19 @@ __vfs_setxattr_locked(struct dentry *dentry, const char *name,
 {
 	struct inode *inode = dentry->d_inode;
 	int error;
-
+#if CONFIG_TRUENAS
+	if (IS_NFSV4ACL(inode)) {
+		error = xattr_permission(inode, name, MAY_WRITE);
+		if (error) {
```

Comment:
Same question as before in `fs/attr.c:314` 😏

Reply:
Same general answer. FreeBSD actually checks only WRITE_DATA. We check ACE4_WRITE_DATA then ACE4_WRITE_NAMED_ATTRS. 

There's a potential performance impact any time we start evaluating permissions that aren't ACE4_READ_DATA, ACE4_WRITE_DATA, and ACE4_EXECUTE. If mask can be represented by POSIX permissions, then we can check the znode pflags for whether the ACL is trivial. If it is trivial, then we know that the permissions can be fully expressed by the POSIX mode and bypass a zfs_access() call in favor of generic_permission().